File: app/server/notify/channels/pushdeer.ts
/**
 * PushDeer 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'pushdeer' });

export const pushdeerChannel: NotifyChannel = {
    name: 'pushdeer',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('DEER_KEY')) {
                resolve({ success: false, message: 'DEER_KEY 未配置' });
                return;
            }

            const DEER_KEY = getConfig('DEER_KEY');
            const DEER_URL = getConfig('DEER_URL') || 'https://api2.pushdeer.com/message/push';

            // PushDeer 建议对消息内容进行 urlencode
            const encodedDesp = encodeURI(desp);
            const body = `pushkey=${DEER_KEY}&text=${text}&desp=${encodedDesp}&type=markdown`;

            $.post({
                url: DEER_URL,
                body,
                headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
            }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'PushDeer 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    const result = data as { content?: { result?: any[] } };
                    if (result?.content?.result && result.content.result.length > 0) {
                        logger.info('PushDeer 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else {
                        logger.warn({ body: data }, 'PushDeer 发送失败');
                        resolve({ success: false, message: '发送失败' });
                    }
                }
            });
        });
    }
};

if (hasConfig('DEER_KEY')) {
    pushdeerChannel.enabled = true;
}

export default pushdeerChannel;

File: app/server/internal/notify/channels/pushdeer.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/url"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type PushDeer struct {
	enabled bool
}

func (c *PushDeer) Name() string            { return "pushdeer" }
func (c *PushDeer) Enabled() bool           { return c.enabled }
func (c *PushDeer) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *PushDeer) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("DEER_KEY") {
		return notify.NotifyResult{Success: false, Message: "DEER_KEY 未配置"}
	}

	deerKey := notify.GetConfig("DEER_KEY")
	deerURL := notify.GetConfig("DEER_URL")
	if deerURL == "" {
		deerURL = "https://api2.pushdeer.com/message/push"
	}

	// PushDeer recommends URL-encoding the content
	encodedDesp := url.QueryEscape(desp)
	body := fmt.Sprintf("pushkey=%s&text=%s&desp=%s&type=markdown", deerKey, text, encodedDesp)

	resp, err := notify.HTTPPost(deerURL, notify.HttpRequestOptions{
		Body:    body,
		Headers: map[string]string{"Content-Type": "application/x-www-form-urlencoded"},
	})
	if err != nil {
		slog.Error("PushDeer 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	var result struct {
		Content *struct {
			Result []interface{} `json:"result"`
		} `json:"content"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &result); err != nil {
		slog.Error("PushDeer 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if result.Content != nil && len(result.Content.Result) > 0 {
		slog.Info("PushDeer 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	slog.Warn("PushDeer 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "发送失败"}
}

func init() {
	ch := &PushDeer{}
	if notify.HasConfig("DEER_KEY") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
